### PR TITLE
GH-2078 property path negation and inversion

### PR DESCRIPTION
**Pull request by hmottestad**
GitHub issue resolved: #2078 

Briefly describe the changes proposed in this PR:

---- 
PR Author Checklist: 

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

File: compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11QueryTest.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import org.eclipse.rdf4j.repository.sail.SailRepository;
+
+import java.net.URL;
+
+public class MemorySPARQL11QueryTest extends SPARQLQueryTest {
+
+	public static Test suite() throws Exception {
+
+		URL manifestUrl = SPARQL11ManifestTest.class.getResource("/testcases-sparql-1.1/manifest-evaluation.ttl");
+
```
Comment (hmottestad) [inline compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11QueryTest.java:24]: @jeenbroekstra I'm wondering about these tests suites.....couldn't find that they were being run anywhere, created my own classes to run them, quite a few of them are failing.....do we run these somewhere else? Or have we abandoned them?

File: core/queryparser/sparql/src/test/java/org/eclipse/rdf4j/query/parser/sparql/SPARQLParserTest.java [hunk 2]
```diff
@@ -283,4 +287,34 @@ public void testLongUnicode() throws Exception {
...
+		TupleExpr tupleExpr = parsedQuery.getTupleExpr();
+
+		Slice slice = (Slice) tupleExpr;
+		Union union = (Union) slice.getArg();
+
+		Var leftSubjectVar = ((StatementPattern) union.getLeftArg()).getSubjectVar();
+		Var rightSubjectVar = ((StatementPattern) ((Filter) union.getRightArg()).getArg()).getSubjectVar();
+
+		assertEquals(leftSubjectVar, rightSubjectVar);
+
```
Comment (hmottestad) [inline core/queryparser/sparql/src/test/java/org/eclipse/rdf4j/query/parser/sparql/SPARQLParserTest.java:317]: Going to need help with fixing the underlying issue for this test.

File: compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11QueryTest.java (hunk 1 above)
Comment (jeenbroekstra) [reply compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11QueryTest.java:24]: I'm not sure I understand. This is a compliance test class, if you run `mvn verify` it gets executed, so these tests are very much part of our suite and are run every time the CI system builds/tests the project.

Comment (jeenbroekstra) [reply compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11QueryTest.java:24]: We have not. They are executed by the classes in package `org.eclipse.rdf4j.query.parser.sparql.manifest`, part of the sparql-compliance module.

Comment (jeenbroekstra) [reply compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11QueryTest.java:24]: Ah, hang on, I see what you mean, you're talking specifically about the `testcases-sparql-1.1` directory, as opposed to `testcases-sparql-1.1-w3c`.

Hm, yes, these look abandoned. From the top of my head these were tests I used while SPARQL was in working draft stage (so before there was a complete official W3C test suite), and I think we initially tried to use it to capture our own additional test cases, but it turned out to be difficult to maintain in parallel with the official W3C suite (and using manifests is cumbersome for running individual tests, as you know). So we switched to using only the official manifest test suite, and for our own additional test cases just adding test cases to the ComplexSPARQLQueryTest class.

Comment (jeenbroekstra) [reply compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11QueryTest.java:24]: I would not worry about these for now - this needs a cleanup and perhaps there are tests in there that we shouldn't abandon, but it's also likely that quite a few of these tests are based on a draft stage of SPARQL, where the later behavior in the standard changed (as a case in point a lot of the property path cases cover things like `ex:prop{1,2}` which was a draft proposal at one point, but never made it into the standard).

File: compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL12QueryTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import junit.framework.Test;
+import org.eclipse.rdf4j.query.Dataset;
+import org.eclipse.rdf4j.query.parser.sparql.manifest.SPARQL11ManifestTest;
+import org.eclipse.rdf4j.query.parser.sparql.manifest.SPARQLQueryTest;
+import org.eclipse.rdf4j.repository.Repository;
+import org.eclipse.rdf4j.repository.sail.SailRepository;
+
+import java.net.URL;
+
+public class MemorySPARQL12QueryTest extends SPARQLQueryTest {
```
Comment (jeenbroekstra) [inline compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL12QueryTest.java:19]: This duplicates `org.eclipse.rdf4j.query.parser.sparql.manifest.SPARQL12QueryTest`.

File: compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11W3CQueryTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import junit.framework.Test;
+import org.eclipse.rdf4j.query.Dataset;
+import org.eclipse.rdf4j.query.parser.sparql.manifest.SPARQL11ManifestTest;
+import org.eclipse.rdf4j.query.parser.sparql.manifest.SPARQLQueryTest;
+import org.eclipse.rdf4j.repository.Repository;
+import org.eclipse.rdf4j.repository.sail.SailRepository;
+
+import java.net.URL;
+
+public class MemorySPARQL11W3CQueryTest extends SPARQLQueryTest {
```
Comment (jeenbroekstra) [inline compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL11W3CQueryTest.java:19]: This duplicates `org.eclipse.rdf4j.query.parser.sparql.manifest.W3CApprovedSPARQL11QueryTest`.

File: compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL10W3CQueryTest.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import junit.framework.Test;
+import org.eclipse.rdf4j.query.Dataset;
+import org.eclipse.rdf4j.query.parser.sparql.manifest.SPARQL11ManifestTest;
+import org.eclipse.rdf4j.query.parser.sparql.manifest.SPARQLQueryTest;
+import org.eclipse.rdf4j.repository.Repository;
+import org.eclipse.rdf4j.repository.sail.SailRepository;
+
+import java.net.URL;
+
+public class MemorySPARQL10W3CQueryTest extends SPARQLQueryTest {
```
Comment (jeenbroekstra) [inline compliance/sparql/src/test/java/org/eclipse/rdf4j/sail/memory/MemorySPARQL10W3CQueryTest.java:19]: This duplicates `org.eclipse.rdf4j.query.parser.sparql.manifest.W3CApprovedSPARQL10QueryTest`.

File: core/queryparser/sparql/src/test/java/org/eclipse/rdf4j/query/parser/sparql/SPARQLParserTest.java (hunk 2 above)
Comment (jeenbroekstra) [reply core/queryparser/sparql/src/test/java/org/eclipse/rdf4j/query/parser/sparql/SPARQLParserTest.java:317]: Yes, unfortunately this code is the worst kind of god-class. I think I see what needs to be fixed, but I'll need a bit of time.

**Comment by jeenbroekstra** [comment]
I've pushed what I think is a decent fix for the immediate issue. 

@hmottestad this is really a bug fix and should be merged to master, but unfortunately the feature branch is based off of develop. Could you rebase and target master?

**Comment by jeenbroekstra** [comment]
ACtually just realized i didn't run all compliance tests - I think I may have overlooked a corner case.

**Comment by hmottestad** [comment]
Thank for fixing this @jeenbroekstra 

I've added a couple more tests, and one of them looks a bit wonky. It's not so much incorrect I think as it is just a bit bloated.

This is the query:
```
PREFIX : <http://example.org/>
ASK { :IBM ^(:|!:) ?jane, ?jane2. }
```

And this is the plan:

```
Slice ( limit=1 )
   Join
      Union
         Join
            StatementPattern (resultSizeEstimate=UNKNOWN)
               Var (name=jane)
               Var (name=_const_7193c6e1_uri, value=http://example.org/, anonymous)
               Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
            StatementPattern (resultSizeEstimate=UNKNOWN)
               Var (name=jane2)
               Var (name=_const_7193c6e1_uri, value=http://example.org/, anonymous)
               Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
         Filter
            Compare (!=)
               Var (name=_anon_ee21cdb5_6a14_4539_a854_ed672dcc589d, anonymous)
               ValueConstant (value=http://example.org/)
            Join
               StatementPattern (resultSizeEstimate=UNKNOWN)
                  Var (name=jane2)
                  Var (name=_anon_ee21cdb5_6a14_4539_a854_ed672dcc589d, anonymous)
                  Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
               StatementPattern (resultSizeEstimate=UNKNOWN)
                  Var (name=jane)
                  Var (name=_anon_ee21cdb5_6a14_4539_a854_ed672dcc589d, anonymous)
                  Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
      Union
         Join
            StatementPattern (resultSizeEstimate=UNKNOWN)
               Var (name=jane)
               Var (name=_const_7193c6e1_uri, value=http://example.org/, anonymous)
               Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
            StatementPattern (resultSizeEstimate=UNKNOWN)
               Var (name=jane2)
               Var (name=_const_7193c6e1_uri, value=http://example.org/, anonymous)
               Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
         Filter
            Compare (!=)
               Var (name=_anon_ee21cdb5_6a14_4539_a854_ed672dcc589d, anonymous)
               ValueConstant (value=http://example.org/)
            Join
               StatementPattern (resultSizeEstimate=UNKNOWN)
                  Var (name=jane2)
                  Var (name=_anon_ee21cdb5_6a14_4539_a854_ed672dcc589d, anonymous)
                  Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
               StatementPattern (resultSizeEstimate=UNKNOWN)
                  Var (name=jane)
                  Var (name=_anon_ee21cdb5_6a14_4539_a854_ed672dcc589d, anonymous)
                  Var (name=_const_1bf5dbd3_uri, value=http://example.org/IBM, anonymous)
```

**Comment by jeenbroekstra** [comment]
Yes, I did notice that too, and it is a bug of some sort (the produced algebra is not incorrect, just more complex than it needs to be - it essentially duplicate the entire tree for both object list arguments). It is not directly related to this particular bug though.

**Comment by jeenbroekstra** [comment]
Huh, look at that, new feature in github I think: you can convert a pull request back to draft (option appeared for me top-right, underneath the list of reviewers).
